### PR TITLE
Create cspr modification form

### PR DESCRIPTION
**Pull request by ghazwarhili**

### CODE REVIEW

**Comment by thangqp** [comment]
Should correct all sonars which are related to your new files: https://sonarcloud.io/project/issues?id=gridsuite_gridstudy-app&pullRequest=2215&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true

**Comment by thangqp** [comment]
To correct Sonar in new code, new file : https://sonarcloud.io/project/issues?id=gridsuite_gridstudy-app&pullRequest=2215&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true

File: src/translations/en.json [hunk 1]
```diff
@@ -778,19 +788,31 @@
     "ShuntCurrentNumberOfSections": "Current number of sections",
     "ShuntMaximumNumberOfSections": "Maximum number of sections",
     "CreateShuntCompensator": "Create shunt compensator (Linear)",
+    "CreateStaticVarCompensator": "Create CSPR",
```
[the remark this answers is not in the document]
Comment (thangqp) [reply src/translations/en.json:791]: mistake in my previous comment, CSPR is translation in french file .fr

File: src/translations/dynamic/network-modifications-locale-en.ts [hunk 2]
```diff
@@ -34,6 +34,7 @@ const network_modifications_locale_en = {
         '{action, select, TRIP {Trip {computedLabel}} LOCKOUT {Lock out {computedLabel}} ENERGISE_END_ONE {Energise {computedLabel} on {energizedEnd}} ENERGISE_END_TWO {Energise {computedLabel} on {energizedEnd}} SWITCH_ON {Switch on {computedLabel}} other {Equipment operating status modification {computedLabel}}}',
     'network_modifications.SHUNT_COMPENSATOR_CREATION': 'Creating shunt compensator {computedLabel}',
     'network_modifications.SHUNT_COMPENSATOR_MODIFICATION': 'Modifying shunt compensator {computedLabel}',
+    'network_modifications.STATIC_VAR_COMPENSATOR_CREATION': 'Creating cspr {computedLabel}',
```
[the remark this answers is not in the document]
Comment (thangqp) [reply src/translations/dynamic/network-modifications-locale-en.ts:37]: mistake in my previous comment, CSPR is translation in french file .fr

File: src/components/dialogs/network-modifications/static-var-compensator/creation/stand-by-Automaton-form.tsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+import { getFloatNumber } from './stand-by-automaton-form-utils';
+
+type FieldKeys =
+    | 'modeAutomaton'
+    | 'standby'
+    | 'lVoltageSetLimit'
+    | 'hVoltageSetLimit'
+    | 'lVoltageThreshold'
+    | 'hVoltageThreshold';
+export const StandbyAutomatonForm = () => {
```
Comment (thangqp) [inline src/components/dialogs/network-modifications/static-var-compensator/creation/stand-by-Automaton-form.tsx:44]: file name is not in convention, should rename to stand-by-automaton-form or better with standby-automaton-form
stand-by-automaton-form-utils can be rename as standby-automaton-form-utils

File: src/components/dialogs/network-modifications/static-var-compensator/creation/set-points-limits-form.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    currentNode: { id: UUID };
+    voltageLevelOptions: {};
+}
+export const SetPointsLimitsForm: FunctionComponent<SetPointsLimitsFormProps> = ({
+    studyUuid,
+    currentNode,
+    voltageLevelOptions,
+}) => {
+    const watchCharacteristicsChoice = useWatch({ name: CHARACTERISTICS_CHOICE });
+    const watchRegulationMode = useWatch({ name: VOLTAGE_REGULATION_MODE });
```
Comment (thangqp) [inline src/components/dialogs/network-modifications/static-var-compensator/creation/set-points-limits-form.tsx:45]:
```suggestion
   // a tricky solution to rerender voltage/reactive setpoints field with label changed between required <-> optional
    useWatch({ name: VOLTAGE_REGULATION_MODE });
```

File: src/components/dialogs/network-modifications/static-var-compensator/creation/static-var-compensator-creation-dialog.tsx [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+        delay: FORM_LOADING_DELAY,
+    });
+
+    const onValidationError = (errors: any) => {
+        console.log('=============================errors', errors);
+        let tabsInError = [];
+        if (errors?.[CONNECTIVITY] !== undefined) {
+            tabsInError.push(StaticVarCompensatorCreationDialogTab.CONNECTIVITY_TAB);
+        }
+        if (errors?.[SETPOINTS_LIMITS] !== undefined) {
```
Comment (thangqp) [inline src/components/dialogs/network-modifications/static-var-compensator/creation/static-var-compensator-creation-dialog.tsx:382]: not present in schema so this tab is not shown in red when some inside fields is on error

File: src/components/dialogs/network-modifications/static-var-compensator/creation/static-var-compensator-creation-dialog.tsx [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+            tabsInError.push(StaticVarCompensatorCreationDialogTab.SET_POINTS_LIMITS_TAB);
+        }
+        if (errors?.[AUTOMATON] !== undefined) {
```
Comment (thangqp) [inline src/components/dialogs/network-modifications/static-var-compensator/creation/static-var-compensator-creation-dialog.tsx:385]: same as setpoints, not present in schema

File: src/components/dialogs/network-modifications/static-var-compensator/creation/static-var-compensator-creation-dialog.tsx [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+            tabsInError.push(StaticVarCompensatorCreationDialogTab.AUTOMATON_TAB);
+        }
+        if (errors?.[ADDIONAL_INFOS] !== undefined) {
```
Comment (thangqp) [inline src/components/dialogs/network-modifications/static-var-compensator/creation/static-var-compensator-creation-dialog.tsx:388]: same as setpoints, not present in schema

File: src/components/dialogs/network-modifications/static-var-compensator/creation/susceptance-area.tsx [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    );
+
+    return (
+        <>
+            {isSusceptance && (
+                <Grid container spacing={2} padding={2}>
+                    {gridItem(minSusceptanceField, 3)}
+                    {gridItem(sliderS, 3)}
+                    {gridItem(maxSusceptanceField, 3)}
+                    {gridItem(susceptanceField, 3)}
```
Comment (thangqp) [inline src/components/dialogs/network-modifications/static-var-compensator/creation/susceptance-area.tsx:141]: this field shows NAN when max and min are not edited in tab setpoints

File: src/components/dialogs/network-modifications/static-var-compensator/creation/susceptance-area.tsx [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+        return value;
+    };
+
+    const handleQnomValueChange = (value: any) => {
+        setValue(SLIDER_Q_NOMINAL, getFloatNumber(value));
+        return value;
+    };
+
+    const minSusceptanceField = (
+        <FloatInput
```
Comment (thangqp) [inline src/components/dialogs/network-modifications/static-var-compensator/creation/susceptance-area.tsx:67]: value should be propagated from the parent component, avoid to use rhf component

File: src/components/dialogs/network-modifications/static-var-compensator/creation/susceptance-area.tsx [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+import { getFloatNumber } from './stand-by-automaton-form-utils';
+import { isValidPercentage } from '../../../percentage-area/percentage-area-utils';
+
+export interface SusceptanceAreaProps {
+    isSusceptance: boolean;
+    isQFixe: boolean;
+    min: number;
+    max: number;
+}
+export const SusceptanceArea: FunctionComponent<SusceptanceAreaProps> = ({ isSusceptance, isQFixe, min, max }) => {
```
Comment (thangqp) [inline src/components/dialogs/network-modifications/static-var-compensator/creation/susceptance-area.tsx:32]: In my opinion, we can make the component more generic with 4 JSX sub-component, min (TextField), max (TextField), slider(SliderInput), input(TextInput) with props {minLabel, maxLabel, sliderLabel, inputLabel, sliderName, inputName, min, max}

File: src/translations/en.json [hunk 11]
```diff
@@ -647,7 +647,7 @@
     "BUSBAR_SECTION": "Bus bar section",
     "BY_FORMULA": "By formula",
     "SHUNT_COMPENSATOR": "Shunt Compensator",
-    "STATIC_VAR_COMPENSATOR": "Static Var Compensator",
+    "STATIC_VAR_COMPENSATOR": "Compensator static var",
```
Comment (thangqp) [inline src/translations/en.json:650]: Static var compensator

to update also in others

File: src/components/dialogs/network-modifications/static-var-compensator/creation/automate-form.tsx [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+    const watchQuNomMin = useWatch({ name: MIN_Q_AT_NOMINAL_V });
+    const watchQuNomMax = useWatch({ name: MAX_Q_AT_NOMINAL_V });
+
+    const [standByDisabled, setStandByDisabled] = useState(watchVoltageMode === VOLTAGE_REGULATION_MODES.VOLTAGE.id);
+    const isSusceptance = watchChoiceAutomate === CHARACTERISTICS_CHOICES.SUSCEPTANCE.id;
+    const isQFixe = watchChoiceAutomate === CHARACTERISTICS_CHOICES.Q_AT_NOMINAL_V.id;
+
+    useEffect(() => {
+        setStandByDisabled(watchVoltageMode === VOLTAGE_REGULATION_MODES.VOLTAGE.id);
+    }, [setStandByDisabled, watchVoltageMode]);
```
[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/dialogs/network-modifications/static-var-compensator/creation/automate-form.tsx]: OK, it is used to set check box standby automaton to false

File: src/components/dialogs/network-modifications/static-var-compensator/creation/set-points-limits-form.tsx [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    voltageLevelOptions: {};
+}
+export const SetPointsLimitsForm: FunctionComponent<SetPointsLimitsFormProps> = ({
+    studyUuid,
+    currentNode,
+    voltageLevelOptions,
+}) => {
+    const [formState, setFormState] = useState({
+        openWatchProps: CHARACTERISTICS_CHOICES.Q_AT_NOMINAL_V.id,
+        openWatchVoltageMode: VOLTAGE_REGULATION_MODES.OFF.id,
```
[the remark this answers is not in the document]
Comment (thangqp) [reply src/components/dialogs/network-modifications/static-var-compensator/creation/set-points-limits-form.tsx]: tricky solution is used to rerender children components to update optional label constraints 
useWatch({ name: `${id}.${VOLTAGE_REGULATION_MODE}` });

File: src/components/dialogs/network-modifications/static-var-compensator/creation/susceptance-area.tsx (hunk 10 above)
Comment (thangqp) [reply src/components/dialogs/network-modifications/static-var-compensator/creation/susceptance-area.tsx]: new solution is useWatch to be reactive with changes from setpoint form

File: src/components/dialogs/network-modifications/static-var-compensator/creation/susceptance-area.tsx [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+            adornment={SusceptanceAdornment}
+            acceptValue={isValidPercentage}
+            outputTransform={handleSusceptanceValueChange}
+        />
+    );
+
+    const qAtNominalVField = (
+        <TextInput
+            name={`${id}.${Q0}`}
+            label="q0"
```
Comment (thangqp) [inline src/components/dialogs/network-modifications/static-var-compensator/creation/susceptance-area.tsx:131]: CSPR_Unit
unit shown here due to a shared translation key which contains unit
I have no solution for that except creating another translation key

File: src/components/dialogs/network-modifications/static-var-compensator/creation/susceptance-area.tsx [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+    const maxQ = useWatch({ name: `${SETPOINTS_LIMITS}.${MAX_Q_AT_NOMINAL_V}` });
+
+    useEffect(() => {
+        let avgSfixeValue = (getFloatNumber(minS) + getFloatNumber(maxS)) / 2;
+        let avgQfixeValue = (getFloatNumber(minQ) + getFloatNumber(maxQ)) / 2;
+        setValue(`${id}.${B0}`, getFloatNumber(avgSfixeValue));
+        setValue(`${id}.${SLIDER_SUSCEPTANCE}`, getFloatNumber(avgSfixeValue));
+        setValue(`${id}.${Q0}`, getFloatNumber(avgQfixeValue));
+        setValue(`${id}.${SLIDER_Q_NOMINAL}`, getFloatNumber(avgQfixeValue));
+    }, [setValue, minS, minQ, maxS, maxQ, id]);
```
Comment (thangqp) [inline src/components/dialogs/network-modifications/static-var-compensator/creation/susceptance-area.tsx:48]: CSPR_restore

BUG, we can not restore b0 or q0 value which is saved in the previous edition.
In my opinion, auto-compute b0 or q0 value should be done only when it has not value yet.

File: src/components/dialogs/network-modifications/static-var-compensator/creation/static-var-compensator-creation-dialog.tsx [hunk 16]
```diff
@@ -0,0 +1,481 @@
...
+                aria-labelledby="dialog-create-staticCompensator"
+                titleId="CreateStaticVarCompensator"
+                subtitle={headerAndTabs}
+                searchCopy={searchCopy}
+                onValidationError={onValidationError}
+                open={open}
+                isDataFetching={isUpdate && editDataFetchStatus === FetchStatus.RUNNING}
+                PaperProps={{
+                    sx: {
+                        height: '55vh', // we want the dialog height to be fixed even when switching tabs
```
Comment (etiennehomer) [inline src/components/dialogs/network-modifications/static-var-compensator/creation/static-var-compensator-creation-dialog.tsx:463]: You should remove the slider we get on a laptop screen for SetPointsLimitsForm and StandbyAutomatonForm

Comment (thangqp) [reply src/components/dialogs/network-modifications/static-var-compensator/creation/static-var-compensator-creation-dialog.tsx:463]: May be set to be larger than 55vh in order to disable the scroller?

Comment (ghazwarhili) [reply src/components/dialogs/network-modifications/static-var-compensator/creation/static-var-compensator-creation-dialog.tsx]: I set the height to '95vh' like the other dialogs

File: src/translations/en.json [hunk 17]
```diff
@@ -1371,9 +1396,13 @@
     "EquipmentSearchPlaceholder": "Search",
     "GeneratedModification": "Generated-modification",
     "guidancePopUp.title": "Selection on the map",
+    "guidancePopUp.firstVariant":"Click on the map to start drawing a polygon.",
```
Comment (etiennehomer) [inline src/translations/en.json:1399]: Merge failure ? To be corrected

File: src/translations/en.json [hunk 18]
```diff
@@ -1172,6 +1196,7 @@
     "menu.createMultiple": "Create multiple equipment",
     "menu.modifyMultiple": "Modify multiple equipment",
     "ShuntCompensator": "Shunt compensator",
+    "StaticVarCompensator": "Static var compensator",
```
Comment (etiennehomer) [inline src/translations/en.json:1199]: we already have : "STATIC_VAR_COMPENSATOR": "Static Var Compensator",

File: src/translations/fr.json [hunk 19]
```diff
@@ -780,19 +792,32 @@
...
     "sectionCount": "Gradin courant",
     "SwitchedOnMaxQAtNominalV": "Q enclenchée à tension nominale",
     "SwitchedOnMaxSusceptance": "Susceptance enclenchée",
-    "maxQAtNominalV": "Q installée à tension nominale",
-    "maxSusceptance": "Susceptance installée",
+    "maxQAtNominalV": "Q max à tension nominale",
+    "minQAtNominalV": "Q min à tension nominale",
+    "maxQ": "Q max",
+    "minQ": "Q min",
+    "maxSusceptance": "Susceptance max",
```
Comment (etiennehomer) [inline src/translations/fr.json:819]:
```suggestion
    "maxSusceptance": "Susceptance maximale",
```

File: src/translations/fr.json [hunk 20]
```diff
@@ -780,19 +792,32 @@
...
+    "minSusceptance": "Susceptance min",
```
Comment (etiennehomer) [inline src/translations/fr.json:820]:
```suggestion
    "minSusceptance": "Susceptance minimale",
```

File: src/translations/fr.json (hunk 19 above)
Comment (etiennehomer) [inline src/translations/fr.json:819]: This gonna change few translations. Is it sure that it is what we want ?

File: src/translations/fr.json [hunk 21]
```diff
@@ -780,19 +792,32 @@
...
     "ModifyShuntCompensator": "Modifier un moyen de compensation statique (Linéaire)",
     "MaximumSectionCountMustBeGreaterOrEqualToOne": "Le nombre de gradins doit être supérieur ou égal à 1",
     "SectionCountMustBeBetweenZeroAndMaximumSectionCount": "La valeur de la prise courante doit être comprise entre 0 et le nombre de gradins",
     "maximumSectionCount": "Nombre de gradins",
     "sectionCount": "Gradin courant",
     "SwitchedOnMaxQAtNominalV": "Q enclenchée à tension nominale",
     "SwitchedOnMaxSusceptance": "Susceptance enclenchée",
-    "maxQAtNominalV": "Q installée à tension nominale",
-    "maxSusceptance": "Susceptance installée",
+    "maxQAtNominalV": "Q max à tension nominale",
```
Comment (etiennehomer) [inline src/translations/fr.json:815]: This gonna change few translations. Is it sure that it is what we want ?

File: src/translations/en.json [hunk 22]
```diff
@@ -140,6 +140,7 @@
     "ReactivePower": "q (MVar)",
     "p0": "Constant p (MW)",
     "q0": "Constant q (MVar)",
+    "q0Fixe": "Constant q",
```
Comment (etiennehomer) [inline src/translations/en.json:143]: See "ConstantQWithoutUnit": "Constant Q",

File: src/translations/en.json [hunk 23]
```diff
@@ -764,7 +767,15 @@
     "Capacitor": "Capacitor",
     "QatNominalV": "Q at nominal voltage",
     "QatNominalVLabel": "Q at nominal voltage (MVar)",
+    "OffRegulationText": "Off",
```
Comment (etiennehomer) [inline src/translations/en.json:770]: See "Off": "Off",

**Review by etiennehomer — CHANGES_REQUESTED** [review]

File: src/translations/dynamic/network-modifications-locale-fr.ts [hunk 24]
```diff
@@ -36,6 +36,7 @@ const network_modifications_locale_fr = {
         "{action, select, TRIP {Déclenchement de {computedLabel}} LOCKOUT {Consignation de {computedLabel}} ENERGISE_END_ONE {Mise sous tension à vide de {computedLabel} depuis {energizedEnd}} ENERGISE_END_TWO {Mise sous tension à vide de {computedLabel} depuis {energizedEnd}} SWITCH_ON {Mise en service de {computedLabel}} other {Modification du statut opérationnel de l'équipement {computedLabel}}}",
     'network_modifications.SHUNT_COMPENSATOR_CREATION': "Création d'un moyen de compensation {computedLabel}",
     'network_modifications.SHUNT_COMPENSATOR_MODIFICATION': "Modification d'un moyen de compensation {computedLabel}",
+    'network_modifications.STATIC_VAR_COMPENSATOR_CREATION': "Création d'un cspr {computedLabel}",
```
Comment (etiennehomer) [inline src/translations/dynamic/network-modifications-locale-fr.ts:39]:
```suggestion
    'network_modifications.STATIC_VAR_COMPENSATOR_CREATION': "Création d'un CSPR {computedLabel}",
```

File: src/components/dialogs/network-modifications/static-var-compensator/creation/standby-automaton-form-utils.ts [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+            }),
+        [LOW_VOLTAGE_SET_POINT]: requiredIfAddStandbyAutomaton(yup.number()),
+        [HIGH_VOLTAGE_SET_POINT]: requiredIfAddStandbyAutomaton(yup.number()),
+        [LOW_VOLTAGE_THRESHOLD]: requiredIfAddStandbyAutomaton(yup.number()),
+        [HIGH_VOLTAGE_THRESHOLD]: requiredIfAddStandbyAutomaton(yup.number()),
+        [CHARACTERISTICS_CHOICE_AUTOMATON]: requiredIfAddStandbyAutomaton(yup.string()),
+        [B0]: requiredIfAddStandbyAutomatonAndSusceptanceChoice(yup.number()),
+        [Q0]: requiredIfAddStandbyAutomatonAndQaTNominalVChoice(yup.number()),
+        [SLIDER_SUSCEPTANCE]: requiredIfAddStandbyAutomatonAndSusceptanceChoice(yup.number()),
+        [SLIDER_Q_NOMINAL]: requiredIfAddStandbyAutomatonAndQaTNominalVChoice(yup.number()),
```
Comment (thangqp) [inline src/components/dialogs/network-modifications/static-var-compensator/creation/standby-automaton-form-utils.ts:83]: We can remove these validations. When switching from Susceptance to Q at Nominal in tab 'Automaton', if min max of Q at Nominal are not set in tab 'Set Points & Limit', slider value will be set to null, even Fixed value is filled with a value, the form can not be validated and the slider can not be changed to red.

**Review by thangqp — APPROVED** [review]

File: src/translations/en.json (hunk 23 above)
Comment (ghazwarhili) [reply src/translations/en.json]: ok

File: src/translations/fr.json (hunk 21 above)
Comment (ghazwarhili) [reply src/translations/fr.json]: done

File: src/translations/fr.json (hunk 20 above)
Comment (ghazwarhili) [reply src/translations/fr.json:816]: the translation is set by the PO

File: src/translations/fr.json [hunk 26]
```diff
@@ -508,7 +508,7 @@
     "Constraint": "Contrainte",
     "EquipmentType": "Type d'ouvrage",
     "selectionType": "Type de sélection",
-
+    "StaticVarCompensators": "CSPR",
```
Comment (etiennehomer) [inline src/translations/fr.json:511]: Remove this trad

**Review by etiennehomer — APPROVED** [review]

**Pull request merged**